core/store/src/db/refcount.rs

```diff
@@ -0,0 +1,138 @@
+use std::cmp::Ordering;
+use std::io::{Cursor, Write};
+
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use rocksdb::compaction_filter::Decision;
+use rocksdb::MergeOperands;
+
+use crate::db::RocksDB;
+use crate::DBCol;
+
+/// Refcounted columns store value with rc.
+///
+/// Write write refcount records, reads merge them.
+/// The merged rc should always be positive, if it's not there is a bug in gc.
+///
+/// Refcount record format:
+/// rc = 0 => empty
+/// rc < 0 => 8 bytes little endian rc
+/// rc > 0 => value followed by 8 bytes little endian rc
+///
+pub(crate) fn merge_refcounted_records(result: &mut Vec<u8>, val: &[u8]) {
+    let (bytes, add_rc) = decode_value_with_rc(val);
+    if add_rc == 0 {
+        return;
+    }
+    let (result_bytes, result_rc) = decode_value_with_rc(result);
+    if result_rc == 0 {
+        result.extend_from_slice(val);
+    } else {
+        let rc = result_rc + add_rc;
+        debug_assert!(result_rc <= 0 || add_rc <= 0 || result_bytes == bytes);
+        match rc.cmp(&0) {
+            Ordering::Less => {
```

Comment:
How can this case happen?

Reply:
Depends on the order of merging - if it happens in order of insertion intermediate results shouldn't go negative, but easier not to think about it

Reply:
Looks like it happens in order of insertion, so it would be safe to debug_assert here